File: week2/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

// Define Coleman-Liau constants
const float c1 = 0.0588;
const float c2 = 0.296;
const float c3 = 15.8;

int colemanLiau(string s);
void printResults(int index);

// Implements Coleman-Liau index to estimate reading level of text supplied by user
int main(void)
{
    // Prompt user for text
    string userText = get_string("Text: ");

    // Print results given calculated Coleman-Liau index
    printResults(colemanLiau(userText));
}

// Calculate Coleman-Liau index
// index = 0.0588 * L - 0.296 * S - 15.8
int colemanLiau(string s)
{
    // Parse quantifiable structure of provided text

    int letterCount = 0;
    // Instantiate word count at 1, as first word does not begin with space
    int wordCount = 1;
    int sentenceCount = 0;

    // Iterate through text
    for (int i = 0, n = strlen(s); i < n; i++)
    {
        // Locate alphabetical chars, incrementing letter count
        if (isalpha(s[i]) != 0)
        {
            letterCount++;
        }
        // Locate breaks between words, incrementing word count
        else if (s[i] == ' ')
        {
            wordCount++;
        }
        // Locate end of sentence punctuation, incrementing sentence count
        else if (s[i] == '.' || s[i] == '?' || s[i] == '!')
        {
            sentenceCount++;
        }
    }

    // Calculate algorithm inputs

    // Average letters per 100 words
    float inputL = (100.0 * letterCount) / wordCount;

    // Average sentences per 100 words
    float inputS = (100.0 * sentenceCount) / wordCount;

    // Calculate and return index, rounded to nearest integer
    return (int) round(c1 * inputL - c2 * inputS - c3);
}

// Given Coleman-Liau index, prints corresponding grade level
void printResults(int index)
{
    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index > 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }
}

File: week6/sentimental-readability/readability.py
def main():
    user_text = obtain_text()
    reading_index = coleman_liau(user_text)
    print_reading_level(reading_index)


def obtain_text():
    """Obtain input text passage from the user"""
    while True:
        user_text = input("Text: ")
        # Ensure text is not blank
        if user_text:
            return user_text
        else:
            print("Please provide some text")


def coleman_liau(user_text):
    """Calculate Coleman-Liau index using user text input"""
    # 0.0588 * L - 0.296 * S - 15.8
    # L = letters/100 words
    # S = sentences/100 words
    letter_count = 0
    word_count = 0
    sentence_count = 0

    # Iterate characters in text
    for char in user_text:
        if char.isalpha():
            letter_count += 1
        elif char == " ":
            word_count += 1
        elif char in [".", "!", "?"]:
            sentence_count += 1
    # Add last word to count
    word_count += 1
    # If single sentence with no ending punctuation, sentence count = 1
    if sentence_count == 0:
        sentence_count = 1

    # Calculate input values and resulting reading index
    colemanL = 100 * letter_count / word_count
    colemanS = 100 * sentence_count / word_count
    return 0.0588 * colemanL - 0.296 * colemanS - 15.8


def print_reading_level(reading_index):
    """Prints reading level given Coleman-Liau index as input"""
    if reading_index >= 16:
        print("Grade 16+")
    elif reading_index < 1:
        print("Before Grade 1")
    else:
        print(f"Grade {round(reading_index)}")


if __name__ == "__main__":
    main()
